### PR TITLE
PoC: Multisocketsniff

### PR DESCRIPTION
**Pull request by polybassa**
Hi, this is a proof of concept. For automotive stuff, it would be cool, if I could sniff on multiple sockets of different types at the same time. 

What do you think about this PoC? It's not optimized and it would require a lot more unit tests. Also I don't know about the performance of this changes.

### CODE REVIEW

File: scapy/sendrecv.py [hunk 1]
```diff
@@ -872,15 +871,22 @@ def sniff(count=0, store=True, offline=None, prn=None, lfilter=None,
...
-    _select = lambda sockets, remain: select_func(sockets, remain)[0]
+    sniff_sockets_t = list(set([type(x) for x in sniff_sockets.keys()]))
+    group_sockets = lambda sockets: [[x for x in sockets.keys()
+                                      if type(x) == sniff_sockets_t[i]]
+                                     for i in range(len(sniff_sockets_t))]
+    select_funcs = [x[0].select for x in group_sockets(sniff_sockets)]
+
+    def _select(sockets, remain):
+        lists = list(select_funcs[i](group_sockets(sockets)[i], remain)[0]
+                     for i in range(len(select_funcs)))
```
Comment (gpotter2) [inline scapy/sendrecv.py]: You can use enumerate on select_funcs.

Also your calling group_sockets on each iteration, that’s super inefficient

File: scapy/sendrecv.py [hunk 2]
```diff
@@ -872,15 +871,22 @@ def sniff(count=0, store=True, offline=None, prn=None, lfilter=None,
...
-    select_func = _main_socket.select
-    # We check that all sockets use the same select(), or raise a warning
-    if not all(select_func == sock.select for sock in sniff_sockets):
-        warning("Warning: inconsistent socket types ! The used select function"
-                "will be the one of the first socket")
-    # Now let's build the select function, used later on
-    _select = lambda sockets, remain: select_func(sockets, remain)[0]
+    sniff_sockets_t = list(set([type(x) for x in sniff_sockets.keys()]))
+    group_sockets = lambda sockets: [[x for x in sockets.keys()
+                                      if type(x) == sniff_sockets_t[i]]
```
Comment (gpotter2) [inline scapy/sendrecv.py]: Use isinstance

File: scapy/sendrecv.py [hunk 3]
```diff
@@ -872,15 +871,22 @@ def sniff(count=0, store=True, offline=None, prn=None, lfilter=None,
...
+                                     for i in range(len(sniff_sockets_t))]
```
Comment (gpotter2) [inline scapy/sendrecv.py]: You can use enumerate

**Comment by polybassa** [comment]
@gpotter2 Thanks for your comments. Sniffing on multiple Sockets of different types was not working before. Python-Can sockets, for example, are implemented in the same style as Windows sockets. With this PR, it should also be possible, to sniff on classical network sockets and on Python-Can or ISOTPSoftSockets at the same time. At least, this would be my goal with this PR.

File: scapy/sendrecv.py (hunk 3 above)
Comment (polybassa) [reply scapy/sendrecv.py]: enumerate will cause this error: 
```
Traceback (most recent call last):
  File "<input>", line 13, in <module>
  File "/home/nils/PycharmProjects/secdev_scapy/scapy/sendrecv.py", line 878, in sniff
    select_funcs = [x[0].select for x in group_sockets(sniff_sockets)]
  File "/home/nils/PycharmProjects/secdev_scapy/scapy/sendrecv.py", line 877, in <lambda>
    for i in enumerate(sniff_sockets_t)]
  File "/home/nils/PycharmProjects/secdev_scapy/scapy/sendrecv.py", line 877, in <listcomp>
    for i in enumerate(sniff_sockets_t)]
  File "/home/nils/PycharmProjects/secdev_scapy/scapy/sendrecv.py", line 876, in <listcomp>
    if isinstance(sniff_sockets_t[i], x)]
TypeError: list indices must be integers or slices, not tuple
```

Comment (polybassa) [reply scapy/sendrecv.py]: I could replace `range(len(` here, but I don't see a way to improve line 880

Comment (gpotter2) [reply scapy/sendrecv.py]: Here’s a syntax example
```
for key, value in enumerate(some_list):
    print(counter, value)
```
It provides both index and key at the same time

**Comment by polybassa** [comment]
Is there a preferred way to simulate client-server communication in unit-tests? I would like to setup a server on localhost during a unit test. My test should connect to a server on localhost and send a couple of messages, while sniff is running on the client socket. 

Another thing: Since select is blocking on linux sockets, this change would introduce some undesired behaviour. If other sockets get ready meanwhile the loop is waiting to return from a select, it could happen, that packets are missed. 
What do you think about this? Am I right with this assumption?

**Comment by gpotter2** [comment]
Normally (especially on Linux), calling `select([a_list_of_sockets]...)` returns the sockets that are ready when one of them gets ready.

Because it doesn’t work on windows, and sometimes differently on some OSes, they is that general `select()` function we use.

What you should do is check that all sockets have the same select function (they are all compatible), and then call one single select with all of them (that is kinda what was already in place)

**Comment by polybassa** [comment]
> Normally (especially on Linux), calling `select([a_list_of_sockets]...)` returns the sockets that are ready when one of them gets ready.
> 
> Because it doesn’t work on windows, and sometimes differently on some OSes, they is that general `select()` function we use.
> 
> What you should do is check that all sockets have the same select function (they are all compatible), and then call one single select with all of them (that is kinda what was already in place)

Yes, this is what I'm doing here

> 
> If you absolutely want to support several sockets types, we could match sockets that have the same `select` together, then implement something based on `SelectableObject` (automaton.py) that would select among the selects (that’s becoming maybe over-complicated, we may just want to stick to compatible sockets)
I agree on that. I prefer a easy solution here.

My main concern about this code here is, that if a socket has a blocking receive with timeout exception, the process will wait, until this exception is fired. All messages, which arrive in the meantime will get invalid timestamps, since a receive is not called immediately after they became available on a socket.  

One solution would be to provide a timeout parameter to all sockets with blocking recv functions, or a sniff with timeout. In this way, sniff can set the recv timeout to a minimum, to get back to the next select function as soon as possible. 
Also a blocking select on linux might cause the same effect.

**Comment by gpotter2** [comment]
This is now pretty out of date.
Unless it's fully reworked, I'd be for closing it

**Comment by polybassa** [comment]
Of course. Just close it. I may start a new attempt sometime in the future 

Mit freundlichen Grüßen 

Nils Weiß

**Comment by gpotter2** [comment]
👍

**Pull request closed without merging**